mysql_tests/test_transaction.py

```diff
@@ -0,0 +1,265 @@
+
+    async with bind.acquire() as conn:
+        tx = await conn.transaction()
+        await u.update(nickname="commit").apply()
+        await tx.commit()
+    assert await get_name() == "commit"
+
+    async with bind.acquire() as conn:
+        tx = await conn.transaction()
+        await u.update(nickname="rollback").apply()
+        assert await get_name() == "rollback"
+        await tx.rollback()
+
+    assert await get_name() == "commit"
+
+    # Neither commit nor rollback, should rollback
+    async with bind.acquire() as conn:
+        await conn.transaction()
+        await u.update(nickname="rollback").apply()
+        assert await get_name() == "rollback"
+
+    assert await get_name() == "commit"
+
+    assert init_size == qsize(bind)
+
+
+async def test_engine(bind):
+    init_size = qsize(bind)
+    u, get_name = await _init(bind)
+
+    assert await get_name() != "commit"
+
+    async with bind.transaction():
+        await u.update(nickname="commit").apply()
+    assert await get_name() == "commit"
+
+    with pytest.raises(ZeroDivisionError):
+        async with bind.transaction():
+            await u.update(nickname="rollback").apply()
+            raise ZeroDivisionError
+    assert await get_name() == "commit"
+    assert init_size == qsize(bind)
+
+
+async def test_begin_failed(bind, mocker):
+    from aiomysql.connection import Connection
+
+    init_size = qsize(bind)
+    mocker.patch("aiomysql.connection.Connection.begin")
+    Connection.begin.side_effect = ZeroDivisionError
+    with pytest.raises(ZeroDivisionError):
+        async with bind.transaction():
+            pass  # pragma: no cover
+    assert init_size == qsize(bind)
+
+
+async def test_commit_failed(bind, mocker):
+    from aiomysql.connection import Connection
+
+    init_size = qsize(bind)
+    mocker.patch("aiomysql.connection.Connection.begin")
+    # noinspection PyUnresolvedReferences,PyProtectedMember
+    Connection.begin.side_effect = ZeroDivisionError
+    with pytest.raises(ZeroDivisionError):
+        async with bind.transaction():
+            pass
+    assert init_size == qsize(bind)
+
+
+async def test_reuse(bind):
+    from aiomysql.connection import Connection
+
+    init_size = qsize(bind)
+    async with db.acquire() as conn:
+        async with db.transaction() as tx:
+            assert tx.connection.raw_connection is conn.raw_connection
+            assert isinstance(tx.raw_transaction, Connection)
+            async with db.transaction() as tx2:
+                assert tx2.connection.raw_connection is conn.raw_connection
+            async with db.transaction(reuse=False) as tx2:
+                assert tx2.connection.raw_connection is not conn.raw_connection
+        async with db.transaction(reuse=False) as tx:
+            assert tx.connection.raw_connection is not conn.raw_connection
+            async with db.transaction() as tx2:
+                assert tx2.connection.raw_connection is tx.connection.raw_connection
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=5587950)